host/xtest/sdp_basic.c

```diff
@@ -0,0 +1,555 @@
+			if (!warm_ns_access)
+				break;
+			fprintf(stderr, "info: NonSec map: bad SDP content\n");
+			warm_ns_access = 0;
+			break;
+		}
+	}
+
+	munmap(data, offset + len);
+	return 0;
+}
+
+static int check_sdp_transformed(struct tee_ctx *ctx,
+			  void *ref, size_t offset, size_t len, void *shm_ref)
+{
+	TEEC_SharedMemory *shm = (TEEC_SharedMemory *)shm_ref;
+	char *bref = (char *)ref;
+	char *data;
+	size_t n;
+
+	(void)ctx;
+
+	data = mmap(NULL, offset + len, PROT_READ, MAP_SHARED,
+					shm->registered_fd, 0);
+	if (data == MAP_FAILED)
+		return BUF_MUST_MAP;
+
+	for (n = 0; n < len; n++) {
+		if (data[offset + n] != (unsigned char)(~(*bref++) + 1)) {
+			if (!warm_ns_access)
+				break;
+			fprintf(stderr, "info: NonSec map: bad SDP content\n");
+			warm_ns_access = 0;
+			break;
+		}
+	}
+
+	munmap(data, offset + len);
+	return 0;
+}
+
+static int check_sdp_dumped(struct tee_ctx *ctx, void *ref, size_t len,
+								void *out)
+{
+	char *bref = (char *)ref;
+	char *data = (char *)out;
+	int err = 0;
+
+	(void)ctx;
+
+	while(len--)
+		if (*data++ != (unsigned char)(~(*bref++) + 1))
+			err++;
+
+	return err;
+}
+
+/* 32kB+1 of random data is sufficient */
+#define RANDOM_BUFFER_SIZE	(32 * 1024 + 1)
+static int get_random_bytes(char *out, size_t len)
+{
+	static char *rand_buf = NULL;
+	static size_t rand_idx = 0;
+	int rc;
+
+	if (!rand_buf) {
+		const char rand_dev[] = "/dev/urandom";
+		int fd;
+
+		rand_buf = malloc(RANDOM_BUFFER_SIZE);
+		if (!rand_buf) {
+			fprintf(stderr, "failed to random buffer memory (%d bytes)\n",
+				RANDOM_BUFFER_SIZE);
+			return -1;
+		}
+
+		fd = open(rand_dev, O_RDONLY);
+		if (fd < 0) {
+			fprintf(stderr, "failed to open %s\n", rand_dev);
+			return -1;
+		}
+
+		rc = read(fd, rand_buf, RANDOM_BUFFER_SIZE);
+		if (rc != RANDOM_BUFFER_SIZE) {
+			fprintf(stderr, "failed to read %d bytes from %s\n",
+				RANDOM_BUFFER_SIZE, rand_dev);
+			return -1;
+		}
+		close(fd);
+	}
+
+	while (len) {
+		size_t t_len = (RANDOM_BUFFER_SIZE < len) ? RANDOM_BUFFER_SIZE : len;
+
+		if ((rand_idx + t_len) > RANDOM_BUFFER_SIZE) {
+			int sz_end = RANDOM_BUFFER_SIZE - rand_idx;
+			int sz_beg = t_len - sz_end;
+
+			memcpy(out, rand_buf + rand_idx, sz_end);
+			memcpy(out + sz_end, rand_buf , sz_beg);
+			rand_idx = sz_beg;
+		} else {
+			memcpy(out, rand_buf + rand_idx, t_len);
+			rand_idx += t_len;
+		}
+		len -= t_len;
+	}
+	return 0;
+}
+
+
+static int sdp_basic_test(size_t size, size_t loop, int ion_heap, int ns_map)
+{
+	struct tee_ctx *ctx = NULL;
+	unsigned char *test_buf = NULL;
+	unsigned char *ref_buf = NULL;
+	void *shm_ref = NULL;
+	unsigned int err = 1;
+	int fd = -1;
+	size_t offset;
+
+	if (!loop) {
+		fprintf(stderr, "Error: null loop value\n");
+		return 1;
+	}
+
+	test_buf = malloc(size);
+	ref_buf = malloc(size);
+	if (!test_buf || !ref_buf) {
+		verbose("failed to allocate memory\n");
+		goto out;
+	}
+
+	/* create secure buffer (allocate 'size' + max 'offset' (255) */
```

Comment:
User defines the size of the buffer to be used for the test.
But to enhance test, the application adds an extra offset so that we test buffer which may not be page or cache line aligned. Hence, the application fills buffer with random data, an use a random byte to set an extra offset inside the allocated buffer when invoking the TA.

This is why we allocate here the expected size + 255 (max value for a random byte), to insure we invoke the TA with the expected buffer size but with random buffer alignment.

It seems I should clarify that in the source code.